python/lsst/daf/butler/core/dataIdPacker.py

```diff
@@ -0,0 +1,464 @@
+        metadata, kwds = cls.configure(self.dimensions)
+        return DataIdPackerFactory(self.dimensions, cls=cls, metadata=metadata, **kwds)
+
+
+class DataIdPackerLibrary:
+    """A manager that manages `DataIdPacker` classes for `Registry.`
+
+    This class is intended for internal use by `Registry` only.
+
+    Parameters
+    ----------
+    universe : `DimensionGraph`
+        All dimensions known to the `Registry`.
+    config : `Config`
+        The "dataIdPackers" subsection of a `RegistryConfig`.
+    """
+
+    def __init__(self, universe, config):
+        self._metadata = DimensionKeyDict(keys=universe.elements, factory=set)
+        self._loaders = defaultdict(dict)
+        self._factories = defaultdict(dict)
+        for name, subconfig in config.items():
+            dimensions = DataIdPackerDimensions.fromConfig(universe.universe, subconfig)
+            clsName = subconfig["cls"]
+            self._loaders[dimensions.given][name] = DataIdPackerLoader(dimensions, clsName)
+
+    def load(self, dimensions):
+        """Import `DataIdPacker` classes for the given dimensions and return a
+        dictionary of metadata fields needed to construct them.
+
+        This must be called explicitly with
+        ``dataId.dimensions(implied=True)`` before any calls to `extract` for
+        a data ID.
+
+        Parameters
+        ----------
+        dimensions : iterable over `Dimension` or `str`
+            Dimensions that must be a superset of the "required" dimensions
+            of any matched `DataIdPacker` classes.
+
+        Returns
+        -------
+        metadata : `~collections.abc.Mapping`
+            A mapping with `DimensionElement` keys and `~collections.abc.Set`
+            of `str` values indicating the dimension metadata fields that must
+            be included in a `DataId` to ensure relevant packers may be
+            constructed.  May (as an optimization) include additional entries
+            not relevant for the given dimensions and/or return an internal
+            dict that should not be modified by the caller.
+
+        Notes
+        -----
+        This is called early in the process of `DataId` expansion to import
+        `DataIdPacker` classes that may be relevant for it; we defer those
+        imports to this point avoid bringing down `Registry` construction just
+        because some packer we weren't actually planning to use in a
+        particular session is in an extension package not in ``sys.path``.
+        Once the packer class is imported, we can retreive the dictionary of
+        metadata entries it needs for instantiation, and hence ensure those
+        entries are always loaded whenever data IDs are expanded.
+        """
+        # Standardize dimension input
+        dimensions = toNameSet(dimensions)
+        matched = []
+        for given, loaders in self._loaders.items():
+            if given.issubset(dimensions):
+                for name, loader in loaders.items():
```

Comment:
it is confusing to have loaders be part of a tuple on items in the dictionary named _loaders